### DISCUSSION TITLE
Send zipped folder as response

### DISCUSSION THREAD

**Discussion by SachinMaharana** [vercel/next.js#30233] [Help]
I want to send a zipped folder as response from an api endpoint. I wonder how would the code would look like. Currently i have this and unfortunately it doesn't work. Any suggestions would be of great help.

```js
const fse = require("fs-extra");
const AdmZip = require("adm-zip");
import path from "path";

const destDir = `dest`;

export default function handler(req, res) {
  let zip = new AdmZip();
  const filePath = path.join(process.cwd(), `/dest`);
  console.log(filePath);
  zip.addLocalFolder(filePath);
  let zipeContents = zip.toBuffer();
  const fileName = "uploads.zip";

  res.setHeader("Content-Type", "application/zip");
  res.setHeader("Content-Disposition", `attachment; filename=${fileName}`);
  res.setHeader("Content-Length", zipeContents.length);
  res.send(zipeContents);

}
```

**Comment by gaston-flores** [comment]
Strange, your example works for me with pretty much no changes at all. 

```js
const AdmZip = require("adm-zip");
const path = require("path");

export default function handler(req, res) {
  let zip = new AdmZip();
  const filePath = path.join(process.cwd(), `/dest`);
  console.log("cwd:", filePath);
  zip.addLocalFolder(filePath);
  let zipeContents = zip.toBuffer();
  const fileName = "uploads.zip";

  res.setHeader("Content-Type", "application/zip");
  res.setHeader("Content-Disposition", `attachment; filename=${fileName}`);
  res.setHeader("Content-Length", zipeContents.length);
  res.send(zipeContents);
}
```

I made a [codesandbox example here.](https://codesandbox.io/s/next-zip-download-xd7qd?file=/pages/api/get-zip.js) You might have to access the URL directly to get the download, but it should work:

![image](https://user-images.githubusercontent.com/18354222/138611260-f2d251c7-b0d1-4ba9-832f-805f4ad18085.png)

What error are you getting exactly?

**Comment by SachinMaharana** [comment]
Thanks for your time @gaston-flores . I did try with your example codesandbox and it certainly works. May i know, how can i trigger from the react code?(i believe my confusion arises from this aspect, and my failure to figure this out.)

Currently, i have a button and on clicking i want to download the zipped folder(after POST request with some payload.)
(on click should i just send a post request and go to /api/get-zip url? )

```js
const handleSubmission = async () => {
    const info = {
      name: "Kieran",
      age: 26,
    };

    const res = await fetch("/api/get-zip", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(info),
    });
   // not sure how i am to handle it here?
    // const data = await res.json();
    // console.log(data); // { message: "success" }
  };
```
```js
  <Button onClick={handleSubmission}>
          Download
</Button>
```

**Reply by gaston-flores** [reply]
You can receive the data from the request as a blob, create an `a` element with the blob as the `href`, and then "click" the link, triggering a download:

```js
  const handleSubmit = async () => {
    const info = {
      name: "Kieran",
      age: 26
    };

    const res = await fetch("/api/get-zip", {
      method: "POST",
      headers: {
        "Content-Type": "application/json"
      },
      body: JSON.stringify(info)
    }).then((res) => res.blob());

    const a = document.createElement("a");
    a.href = URL.createObjectURL(res);
    a.setAttribute("download", "uploads.zip");
    a.click();
  };
```

[I updated my example to include this](https://codesandbox.io/s/next-zip-download-xd7qd?file=/pages/index.js).

**Reply by SachinMaharana** [reply]
Thanks a lot @gaston-flores for teaching me this. It's working as expected and learned the technique as well.
